scsi.c

```diff
@@ -0,0 +1,48 @@
+#include <xboxkrnl/xboxkrnl.h>
+
+#include "scsi.h"
+
+static PDEVICE_OBJECT get_dvd_device_object(void)
+{
+    static PDEVICE_OBJECT device = NULL;
+
+    if (device == NULL) {
+	      ANSI_STRING cdrom;  
+        RtlInitAnsiString(&cdrom, "\\Device\\Cdrom0");
+
+        // Get a reference to the dvd object so that we can query it for info.
+        NTSTATUS status = ObReferenceObjectByName(&cdrom, 0, &IoDeviceObjectType, NULL, (void**)&device);
+
+	      if (!NT_SUCCESS(status)) {
+            return NULL;
+        }
+
+        assert(device != NULL);
+    }
+
+    return device;
+}
+
+NTSTATUS scsi_dvd_command(UCHAR data_in, const void* command, size_t command_length, const void* buffer, size_t buffer_length)
+{
+    PDEVICE_OBJECT device = get_dvd_device_object();
+    assert(device != NULL);
+
+    SCSI_PASS_THROUGH_DIRECT pass_through;
+    RtlZeroMemory(&pass_through, sizeof(SCSI_PASS_THROUGH_DIRECT));
+
+    pass_through.Length = sizeof(SCSI_PASS_THROUGH_DIRECT);
+    pass_through.DataIn = data_in;
+    pass_through.DataBuffer = buffer;
+    pass_through.DataTransferLength = buffer_length;
+
+    assert(command_length <= sizeof(pass_through.Cdb));
+    memcpy(&pass_through.Cdb, command, command_length);
+
+    NTSTATUS status = IoSynchronousDeviceIoControlRequest(IOCTL_SCSI_PASS_THROUGH_DIRECT,
```

Comment:
It's a bit more than a wrapper lol. IDE is complicated enough as is, I'd rather not write a custom driver. Note that you can also DMA transfers etc.

However, something I'd like to see instead is this functionality being moved into generic kernel calls and python functions. There is almost no benefit having it as a seperate command.

The reason for working on this is that I currently try to get the XGD authentication in XQEMU right. However, the hashes for my kreon dumped DVDs don't match, so I want to see if the xbox returns different data.
In general, forwarding such commands to the original drive while testing would be nice.